File: GameOS/simulation/scheduler.py
"""
GameOS - Python Scheduler (Round Robin + Priority)

Tasarim Karari: Python'da da scheduler yazildi cunku
C scheduler'i ctypes ile baglanirken bu dosya
bagimsiz olarak da kullanilabilir. Ayrica baseline (basit RR)
ve gelismis (Priority RR) karsilastirmasi burada yapilir.
"""

from collections import deque
from process_manager import ProcessState
from logger import logger


class RoundRobinScheduler:
    """Basit Round Robin Scheduler (Baseline)"""

    def __init__(self, time_quantum=3):
        self.time_quantum = time_quantum
        self.ready_queue = deque()
        self.current_time = 0
        self.completed = []
        self.timeline = []  # Gantt chart icin
        logger.info("SCHEDULER",
                    f"Round Robin Scheduler baslatildi | Quantum={time_quantum}")

    def add_process(self, pcb):
        """Process'i ready queue'ya ekle"""
        pcb.state = ProcessState.READY
        self.ready_queue.append(pcb)
        logger.debug("SCHEDULER",
                     f"PID={pcb.pid} ({pcb.name}) kuyruga eklendi")

    def run(self):
        """Round Robin scheduling calistir"""
        logger.info("SCHEDULER", "=== Round Robin Scheduling BASLADI ===")

        while self.ready_queue:
            pcb = self.ready_queue.popleft()
            pcb.state = ProcessState.RUNNING

            # Bu process'in calisacagi sure
            exec_time = min(pcb.remaining_time, self.time_quantum)

            logger.info("SCHEDULER",
                       f"[T={self.current_time:3d}] PID={pcb.pid} ({pcb.name}) "
                       f"calistiriliyor | {exec_time} birim | "
                       f"Kalan: {pcb.remaining_time} -> {pcb.remaining_time - exec_time}")

            # Timeline kaydi
            self.timeline.append({
                "pid": pcb.pid,
                "name": pcb.name,
                "start": self.current_time,
                "duration": exec_time
            })

            # Diger process'lerin bekleme suresini guncelle
            for other in self.ready_queue:
                other.wait_time += exec_time

            pcb.remaining_time -= exec_time
            self.current_time += exec_time

            if pcb.remaining_time <= 0:
                pcb.state = ProcessState.TERMINATED
                pcb.turnaround_time = self.current_time
                self.completed.append(pcb)
                logger.info("SCHEDULER",
                           f"[T={self.current_time:3d}] PID={pcb.pid} ({pcb.name}) "
                           f"TAMAMLANDI | TA={pcb.turnaround_time} | "
                           f"Wait={pcb.wait_time}")
            else:
                pcb.state = ProcessState.READY
                self.ready_queue.append(pcb)

        logger.info("SCHEDULER", "=== Round Robin Scheduling TAMAMLANDI ===")
        return self.completed

    def print_stats(self):
        """Istatistikleri yazdir"""
        if not self.completed:
            print("Henuz tamamlanan process yok.")
            return

        print(f"\n{'='*65}")
        print("Round Robin Scheduler Istatistikleri")
        print(f"{'='*65}")
        print(f"{'PID':<6} {'Ad':<20} {'Burst':<8} {'Bekleme':<10} {'Turnaround':<12}")
        print(f"{'-'*65}")

        total_wait = 0
        total_ta = 0

        for pcb in self.completed:
            print(f"{pcb.pid:<6} {pcb.name:<20} {pcb.burst_time:<8} "
                  f"{pcb.wait_time:<10} {pcb.turnaround_time:<12}")
            total_wait += pcb.wait_time
            total_ta += pcb.turnaround_time

        n = len(self.completed)
        print(f"{'-'*65}")
        print(f"Ortalama Bekleme:    {total_wait / n:.2f}")
        print(f"Ortalama Turnaround: {total_ta / n:.2f}")
        print(f"{'='*65}")

    def print_gantt_chart(self):
        """Basit Gantt chart goster"""
        print(f"\n--- Gantt Chart ---")
        for entry in self.timeline:
            bar = "#" * entry["duration"]
            print(f"T={entry['start']:3d} | PID={entry['pid']} "
                  f"({entry['name']:<15s}) |{bar}| +{entry['duration']}")


class PriorityRoundRobinScheduler(RoundRobinScheduler):
    """
    Gelismis Scheduler: Priority + Round Robin
    Ayni oncelikteki process'ler RR ile, farkli oncelikler
    priority sirasina gore calistirilir.

    Oyun konsolunda: GAME > AUDIO > UI > NETWORK > SAVE
    """

    def __init__(self, time_quantum=3):
        super().__init__(time_quantum)
        logger.info("SCHEDULER",
                    "Priority RR Scheduler baslatildi (gelismis)")

    def run(self):
        """Priority-based Round Robin scheduling"""
        logger.info("SCHEDULER", "=== Priority RR Scheduling BASLADI ===")

        all_processes = list(self.ready_queue)
        self.ready_queue.clear()

        while all_processes:
            # En yuksek oncelikli (en dusuk sayi) process'i bul
            min_priority = min(p.priority for p in all_processes)

            # Bu oncelikteki process'leri RR ile calistir
            current_batch = [p for p in all_processes if p.priority == min_priority]
            remaining = [p for p in all_processes if p.priority != min_priority]

            batch_queue = deque(current_batch)

            while batch_queue:
                pcb = batch_queue.popleft()
                pcb.state = ProcessState.RUNNING

                exec_time = min(pcb.remaining_time, self.time_quantum)

                logger.info("SCHEDULER",
                           f"[T={self.current_time:3d}] PID={pcb.pid} ({pcb.name}) "
                           f"[P{pcb.priority}] | {exec_time} birim")

                self.timeline.append({
                    "pid": pcb.pid,
                    "name": pcb.name,
                    "start": self.current_time,
                    "duration": exec_time
                })

                # Tum bekleyenlerin wait_time guncelle
                for other in batch_queue:
                    other.wait_time += exec_time
                for other in remaining:
                    other.wait_time += exec_time

                pcb.remaining_time -= exec_time
                self.current_time += exec_time

                if pcb.remaining_time <= 0:
                    pcb.state = ProcessState.TERMINATED
                    pcb.turnaround_time = self.current_time
                    self.completed.append(pcb)
                    logger.info("SCHEDULER",
                               f"[T={self.current_time:3d}] PID={pcb.pid} TAMAMLANDI")
                else:
                    pcb.state = ProcessState.READY
                    batch_queue.append(pcb)

            all_processes = remaining

        logger.info("SCHEDULER", "=== Priority RR Scheduling TAMAMLANDI ===")
        return self.completed

File: GameOS/core/scheduler.c
/*
 * GameOS - Oyun Konsolu Isletim Sistemi Simulasyonu
 * Round Robin Scheduler (C implementasyonu)
 *
 * Tasarim Karari: Round Robin secildi cunku oyun konsollarinda
 * birden fazla gorev (oyun, ses, ag, UI) esit CPU zamani almali.
 * Sabit time quantum ile her process adil sekilde calisir.
 */

#include "scheduler.h"

void init_scheduler(ReadyQueue *queue, int quantum) {
    queue->count = 0;
    queue->front = 0;
    queue->rear = -1;
    queue->time_quantum = (quantum > 0) ? quantum : DEFAULT_QUANTUM;
    queue->current_time = 0;
    printf("[SCHEDULER] Baslatildi | Quantum: %d | Max Process: %d\n",
           queue->time_quantum, MAX_PROCESSES);
}

int add_process(ReadyQueue *queue, PCB process) {
    if (queue->count >= MAX_PROCESSES) {
        printf("[SCHEDULER] HATA: Kuyruk dolu! PID %d eklenemedi.\n", process.pid);
        return -1;
    }
    queue->rear = (queue->rear + 1) % MAX_PROCESSES;
    process.state = READY;
    process.wait_time = 0;
    process.turnaround_time = 0;
    queue->processes[queue->rear] = process;
    queue->count++;
    printf("[SCHEDULER] Process eklendi: PID=%d, Ad=%s, Burst=%d, Oncelik=%d\n",
           process.pid, process.name, process.total_time, process.priority);
    return 0;
}

PCB* get_next_process(ReadyQueue *queue) {
    if (queue->count == 0) return NULL;

    // READY durumundaki ilk process'i bul
    for (int i = 0; i < queue->count; i++) {
        int idx = (queue->front + i) % MAX_PROCESSES;
        if (queue->processes[idx].state == READY) {
            return &queue->processes[idx];
        }
    }
    return NULL;
}

void run_round_robin(ReadyQueue *queue) {
    printf("\n========== ROUND ROBIN SCHEDULING BASLADI ==========\n");
    printf("Time Quantum: %d birim\n\n", queue->time_quantum);

    int completed = 0;
    int total = queue->count;
    queue->current_time = 0;

    while (completed < total) {
        int found = 0;

        for (int i = 0; i < total; i++) {
            PCB *p = &queue->processes[i];

            if (p->state == TERMINATED) continue;

            found = 1;
            p->state = RUNNING;

            int exec_time = (p->remaining_time < queue->time_quantum)
                            ? p->remaining_time
                            : queue->time_quantum;

            printf("[T=%3d] PID=%d (%s) calistiriliyor | %d birim | Kalan: %d -> %d\n",
                   queue->current_time, p->pid, p->name,
                   exec_time, p->remaining_time, p->remaining_time - exec_time);

            p->remaining_time -= exec_time;
            queue->current_time += exec_time;

            // Diger READY process'lerin bekleme suresini guncelle
            for (int j = 0; j < total; j++) {
                if (j != i && queue->processes[j].state != TERMINATED) {
                    queue->processes[j].wait_time += exec_time;
                }
            }

            if (p->remaining_time <= 0) {
                p->state = TERMINATED;
                p->turnaround_time = queue->current_time;
                completed++;
                printf("[T=%3d] PID=%d (%s) TAMAMLANDI | Turnaround: %d | Bekleme: %d\n",
                       queue->current_time, p->pid, p->name,
                       p->turnaround_time, p->wait_time);
            } else {
                p->state = READY;
            }
        }

        if (!found) break;
    }

    printf("\n========== SCHEDULING TAMAMLANDI ==========\n");
    print_schedule_stats(queue);
}

void run_priority_rr(ReadyQueue *queue) {
    /*
     * Gelismis Scheduler: Priority + Round Robin
     * Oncelik sirasiyla calistirir, ayni onceliktekiler RR ile
     * Bu, oyun konsolunda oyun process'ine ses/ag'dan daha fazla
     * oncelik vermemizi saglar.
     */
    printf("\n========== PRIORITY ROUND ROBIN SCHEDULING BASLADI ==========\n");
    printf("Time Quantum: %d birim\n\n", queue->time_quantum);

    int completed = 0;
    int total = queue->count;
    queue->current_time = 0;

    while (completed < total) {
        // En yuksek oncelikli (en dusuk sayi) READY process'i bul
        int best_priority = 999;
        for (int i = 0; i < total; i++) {
            if (queue->processes[i].state != TERMINATED &&
                queue->processes[i].priority < best_priority) {
                best_priority = queue->processes[i].priority;
            }
        }

        if (best_priority == 999) break;

        int ran_any = 0;
        for (int i = 0; i < total; i++) {
            PCB *p = &queue->processes[i];

            if (p->state == TERMINATED || p->priority != best_priority)
                continue;

            ran_any = 1;
            p->state = RUNNING;

            int exec_time = (p->remaining_time < queue->time_quantum)
                            ? p->remaining_time
                            : queue->time_quantum;

            printf("[T=%3d] PID=%d (%s) [P%d] calistiriliyor | %d birim | Kalan: %d\n",
                   queue->current_time, p->pid, p->name, p->priority,
                   exec_time, p->remaining_time - exec_time);

            p->remaining_time -= exec_time;
            queue->current_time += exec_time;

            for (int j = 0; j < total; j++) {
                if (j != i && queue->processes[j].state != TERMINATED) {
                    queue->processes[j].wait_time += exec_time;
                }
            }

            if (p->remaining_time <= 0) {
                p->state = TERMINATED;
                p->turnaround_time = queue->current_time;
                completed++;
                printf("[T=%3d] PID=%d (%s) TAMAMLANDI\n",
                       queue->current_time, p->pid, p->name);
            } else {
                p->state = READY;
            }
        }

        if (!ran_any) break;
    }

    printf("\n========== PRIORITY RR TAMAMLANDI ==========\n");
    print_schedule_stats(queue);
}

void print_schedule_stats(ReadyQueue *queue) {
    printf("\n--- Istatistikler ---\n");
    printf("%-6s %-20s %-8s %-12s %-12s\n",
           "PID", "Ad", "Burst", "Bekleme", "Turnaround");
    printf("--------------------------------------------------------------\n");

    float total_wait = 0, total_ta = 0;
    int total = queue->count;

    for (int i = 0; i < total; i++) {
        PCB *p = &queue->processes[i];
        printf("%-6d %-20s %-8d %-12d %-12d\n",
               p->pid, p->name, p->total_time,
               p->wait_time, p->turnaround_time);
        total_wait += p->wait_time;
        total_ta += p->turnaround_time;
    }

    printf("--------------------------------------------------------------\n");
    printf("Ortalama Bekleme Suresi:    %.2f\n", total_wait / total);
    printf("Ortalama Turnaround Suresi: %.2f\n", total_ta / total);
}

/* ====== TEST MAIN ====== */
#ifdef SCHEDULER_TEST

static PCB create_pcb(int pid, const char *name, int priority,
                       int burst, int mem_req) {
    PCB p;
    p.pid = pid;
    strncpy(p.name, name, 63);
    p.name[63] = '\0';
    p.priority = priority;
    p.state = READY;
    p.remaining_time = burst;
    p.total_time = burst;
    p.wait_time = 0;
    p.turnaround_time = 0;
    p.memory_required = mem_req;
    return p;
}

int main(void) {
    printf("╔══════════════════════════════════════════════╗\n");
    printf("║   GameOS Scheduler Test (C)                 ║\n");
    printf("╚══════════════════════════════════════════════╝\n");

    /* --- Oyun konsolu process'leri --- */
    PCB game   = create_pcb(1, "SuperMario",    0, 12, 65536);
    PCB audio  = create_pcb(2, "AudioEngine",   1,  6, 16384);
    PCB net    = create_pcb(3, "OnlineService", 2,  8,  8192);
    PCB ui     = create_pcb(4, "GameUI",        1,  4, 32768);
    PCB save   = create_pcb(5, "AutoSave",      3,  3,  4096);

    /* ---- TEST 1: Basit Round Robin ---- */
    printf("\n\n>>> TEST 1: BASELINE - Round Robin (Quantum=3) <<<\n");
    ReadyQueue rr_queue;
    init_scheduler(&rr_queue, 3);

    add_process(&rr_queue, game);
    add_process(&rr_queue, audio);
    add_process(&rr_queue, net);
    add_process(&rr_queue, ui);
    add_process(&rr_queue, save);

    run_round_robin(&rr_queue);

    /* ---- TEST 2: Priority Round Robin ---- */
    printf("\n\n>>> TEST 2: GELISMIS - Priority Round Robin (Quantum=3) <<<\n");
    ReadyQueue prr_queue;
    init_scheduler(&prr_queue, 3);

    /* Process'leri tekrar olustur (state sifirlanmali) */
    add_process(&prr_queue, create_pcb(1, "SuperMario",    0, 12, 65536));
    add_process(&prr_queue, create_pcb(2, "AudioEngine",   1,  6, 16384));
    add_process(&prr_queue, create_pcb(3, "OnlineService", 2,  8,  8192));
    add_process(&prr_queue, create_pcb(4, "GameUI",        1,  4, 32768));
    add_process(&prr_queue, create_pcb(5, "AutoSave",      3,  3,  4096));

    run_priority_rr(&prr_queue);

    /* ---- TEST 3: Karsilastirma ---- */
    printf("\n\n>>> BASELINE vs GELISMIS KARSILASTIRMASI <<<\n");
    printf("%-25s %-15s %-15s\n", "", "Round Robin", "Priority RR");
    printf("-----------------------------------------------------------\n");
    for (int i = 0; i < rr_queue.count; i++) {
        printf("%-25s Bek=%-5d TA=%-5d   Bek=%-5d TA=%-5d\n",
               rr_queue.processes[i].name,
               rr_queue.processes[i].wait_time,
               rr_queue.processes[i].turnaround_time,
               prr_queue.processes[i].wait_time,
               prr_queue.processes[i].turnaround_time);
    }

    printf("\nSonuc: Priority RR, yuksek oncelikli process'lere (GAME)\n");
    printf("daha dusuk bekleme suresi saglar. Oyun konsolunda render\n");
    printf("islemi oncelikli olmalidir.\n");

    return 0;
}

#endif
